samples/data-sealing/common/CMakeLists.txt

```diff
@@ -0,0 +1,12 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# Use the edger8r to generate C bindings from the EDL file.
+add_custom_command(OUTPUT datasealing_t.h datasealing_t.c
+  COMMAND openenclave::oeedger8r --trusted ${CMAKE_SOURCE_DIR}/datasealing.edl)
+
+# Create a library common to each of our three enclaves.
+add_library(common OBJECT dispatcher.cpp keys.cpp ${CMAKE_CURRENT_BINARY_DIR}/datasealing_t.c)
```

Comment:
@anakrish I could use your help here. It's bizarre, but when trying to change this from OBJECT to STATIC, the enclaves can't link correctly. It's been suggested by @CodeMonkeyLeet and @EMumau that maybe there's a circular dependency, but I'm not sure why that would be.

Reply:
Does adding STATIC still preserve the -fPIC compiler flag?

Reply:
Not sure if it did or not, but I forced it to and still the same issue:

```
> ninja -v -j1
[1/15] cd /home/andschwa/openenclave/samples/data-sealing/build/common && /home/andschwa/openenclave-install/bin/oeedger8r --trusted /home/andschwa/openenclave/samples/data-sealing/datasealing.edl
Generating edge routines for the Open Enclave SDK.
Success.
[2/15] /usr/bin/clang++-7  -DOE_BUILD_ENCLAVE -DOE_USE_LIBSGX -I../ -Icommon -isystem /home/andschwa/openenclave-install/include -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libcxx -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libc -fPIC   -m64 -fPIC -nostdinc -fno-stack-protector -fvisibility=hidden -nostdinc++ -std=gnu++11 -MD -MT common/CMakeFiles/common.dir/dispatcher.cpp.o -MF common/CMakeFiles/common.dir/dispatcher.cpp.o.d -o common/CMakeFiles/common.dir/dispatcher.cpp.o -c ../common/dispatcher.cpp
[3/15] /usr/bin/clang++-7  -DOE_BUILD_ENCLAVE -DOE_USE_LIBSGX -I../ -Icommon -isystem /home/andschwa/openenclave-install/include -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libcxx -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libc -fPIC   -m64 -fPIC -nostdinc -fno-stack-protector -fvisibility=hidden -nostdinc++ -std=gnu++11 -MD -MT common/CMakeFiles/common.dir/keys.cpp.o -MF common/CMakeFiles/common.dir/keys.cpp.o.d -o common/CMakeFiles/common.dir/keys.cpp.o -c ../common/keys.cpp
[4/15] /usr/bin/clang-7 -DOE_BUILD_ENCLAVE -DOE_USE_LIBSGX -I../ -Icommon -isystem /home/andschwa/openenclave-install/include -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libc -fPIC   -m64 -fPIC -nostdinc -fno-stack-protector -fvisibility=hidden -MD -MT common/CMakeFiles/common.dir/datasealing_t.c.o -MF common/CMakeFiles/common.dir/datasealing_t.c.o.d -o common/CMakeFiles/common.dir/datasealing_t.c.o   -c common/datasealing_t.c
[5/15] : && /usr/local/cmake-3.13.1-Linux-x86_64/bin/cmake -E remove common/libcommon.a && /usr/bin/ar qc common/libcommon.a  common/CMakeFiles/common.dir/dispatcher.cpp.o common/CMakeFiles/common.dir/keys.cpp.o common/CMakeFiles/common.dir/datasealing_t.c.o && /usr/bin/ranlib common/libcommon.a && :
[6/15] /usr/bin/clang++-7  -DOE_BUILD_ENCLAVE -DOE_USE_LIBSGX -I../ -Icommon -isystem /home/andschwa/openenclave-install/include -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libcxx -isystem /home/andschwa/openenclave-install/include/openenclave/3rdparty/libc -m64 -fPIC -nostdinc -fno-stack-protector -fvisibility=hidden -nostdinc++ -MD -MT enc1/CMakeFiles/enclave_a_v1.dir/ecalls.cpp.o -MF enc1/CMakeFiles/enclave_a_v1.dir/ecalls.cpp.o.d -o enc1/CMakeFiles/enclave_a_v1.dir/ecalls.cpp.o -c ../enc1/ecalls.cpp
[7/15] : && /usr/bin/clang++-7     enc1/CMakeFiles/enclave_a_v1.dir/ecalls.cpp.o  -o enc1/enclave_a_v1  common/libcommon.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboeenclave.a /home/andschwa/openenclave-install/lib/openenclave/enclave/libmbedtls.a /home/andschwa/openenclave-install/lib/openenclave/enclave/libmbedx509.a /home/andschwa/openenclave-install/lib/openenclave/enclave/libmbedcrypto.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboelibcxx.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboelibc.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboecore.a -nostdlib -nodefaultlibs -nostartfiles -Wl,--no-undefined,-Bstatic,-Bsymbolic,--export-dynamic,-pie,--build-id && :
FAILED: enc1/enclave_a_v1 
: && /usr/bin/clang++-7     enc1/CMakeFiles/enclave_a_v1.dir/ecalls.cpp.o  -o enc1/enclave_a_v1  common/libcommon.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboeenclave.a /home/andschwa/openenclave-install/lib/openenclave/enclave/libmbedtls.a /home/andschwa/openenclave-install/lib/openenclave/enclave/libmbedx509.a /home/andschwa/openenclave-install/lib/openenclave/enclave/libmbedcrypto.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboelibcxx.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboelibc.a /home/andschwa/openenclave-install/lib/openenclave/enclave/liboecore.a -nostdlib -nodefaultlibs -nostartfiles -Wl,--no-undefined,-Bstatic,-Bsymbolic,--export-dynamic,-pie,--build-id && :
/home/andschwa/openenclave-install/lib/openenclave/enclave/liboecore.a(calls.c.o): In function `_handle_call_enclave_function':
/home/andschwa/openenclave/build/package/../../enclave/core/sgx/calls.c:(.text+0x397d): undefined reference to `__oe_ecalls_table_size'
/home/andschwa/openenclave/build/package/../../enclave/core/sgx/calls.c:(.text+0x3ab6): undefined reference to `__oe_ecalls_table'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

Reply:
The issue is due to the linker line.

All .o files specified as part of the linker line will be included; while only needed objects in specified libraries (libcommon.a) will be included. 

To make the link work, there are two options:
1. Use whole-archive to link libcommon.a. 

`-Wl,--whole-archive libcommon.a -Wl,--no-whole-archive`

This has the same effect as if all the object files contained in libcommon.a were explicitly specified in the command line.

2. Specify libcommon.a after liboecore.a in the linker line.

Reply:
THANK YOU